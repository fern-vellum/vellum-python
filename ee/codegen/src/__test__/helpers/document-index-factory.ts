export const mockDocumentIndexFactory = () => ({
  id: "b6ef8809-346e-469c-beed-2e5c4e9844c5",
  name: "my-sweet-document",
});
