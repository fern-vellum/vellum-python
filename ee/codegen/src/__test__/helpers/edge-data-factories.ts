import { WorkflowEdge, WorkflowNode } from "src/types/vellum";

const getSourceHandleId = (
  node: WorkflowNode | [WorkflowNode, string]
): string => {
  if (Array.isArray(node)) {
    const [actualNode, portName] = node;

    if (actualNode.type === "GENERIC") {
      const id = actualNode.ports.find((port) => port.name === portName)?.id;
      if (!id) {
        throw new Error(`Port not found: ${portName}`);
      }

      return id;
    }

    if (actualNode.type === "CONDITIONAL") {
      const id = actualNode.data.conditions[Number(portName)]?.sourceHandleId;
      if (!id) {
        throw new Error("Conditional node has no source handle id");
      }

      return id;
    }

    return getSourceHandleId(actualNode);
  }

  if (node.type === "GENERIC") {
    const port = node.ports[0];
    if (!port) {
      throw new Error("Generic node has no ports");
    }

    return port.id;
  }

  if (
    node.type == "TERMINAL" ||
    node.type === "NOTE" ||
    node.type === "ERROR"
  ) {
    throw new Error(`${node.type} nodes have no source handle id`);
  }

  if (node.type === "CONDITIONAL") {
    throw new Error("Conditional nodes must include a port name");
  }

  return node.data.sourceHandleId;
};

const getTargetHandleId = (
  node: WorkflowNode | [WorkflowNode, number]
): string => {
  if (Array.isArray(node)) {
    const [actualNode, targetIndex] = node;
    if (actualNode.type === "MERGE") {
      const id = actualNode.data.targetHandles[targetIndex]?.id;
      if (!id) {
        throw new Error(`Target handle not found: ${targetIndex}`);
      }

      return id;
    }

    return getTargetHandleId(actualNode);
  }

  if (node.type === "GENERIC") {
    return node.trigger.id;
  }

  if (node.type === "MERGE") {
    throw new Error("Merge nodes must include target handle index");
  }

  if (node.type === "ENTRYPOINT" || node.type === "NOTE") {
    throw new Error(`${node.type} nodes have no target handle id`);
  }

  return node.data.targetHandleId;
};

export type EdgeFactoryNodePair = [
  WorkflowNode | [WorkflowNode, string],
  WorkflowNode | [WorkflowNode, number]
];

export function edgesFactory(nodePairs: EdgeFactoryNodePair[]): WorkflowEdge[] {
  return nodePairs.map(([sourceNode, targetNode], index) => ({
    id: `edge-${index + 1}`,
    type: "DEFAULT",
    sourceNodeId: Array.isArray(sourceNode) ? sourceNode[0].id : sourceNode.id,
    sourceHandleId: getSourceHandleId(sourceNode),
    targetNodeId: Array.isArray(targetNode) ? targetNode[0].id : targetNode.id,
    targetHandleId: getTargetHandleId(targetNode),
  }));
}
