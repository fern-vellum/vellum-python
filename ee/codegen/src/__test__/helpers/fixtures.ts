import { readdirSync } from "fs";
import { basename, join, relative } from "path";

const fixtureDir = join(
  __dirname,
  "..",
  "..",
  "..",
  "..",
  "tests",
  "codegen_integration",
  "fixtures"
);

const allFixtures = readdirSync(fixtureDir);

function getFixtures(
  excludeFixtures: Set<string> | string[] = new Set(),
  includeFixtures: Set<string> | string[] = new Set()
): string[] {
  const excludeSet = Array.isArray(excludeFixtures)
    ? new Set(excludeFixtures)
    : excludeFixtures;
  const includeSet = Array.isArray(includeFixtures)
    ? new Set(includeFixtures)
    : includeFixtures;

  return allFixtures.filter((f) => {
    const fileName = basename(f);

    return (
      (!excludeSet.size || !excludeSet.has(fileName)) &&
      (!includeSet.size || includeSet.has(fileName))
    );
  });
}

function getFixturePaths(root: string) {
  const displayFile = join(
    fixtureDir,
    root,
    "display_data",
    `${basename(root)}.json`
  );
  const codeDir = join(fixtureDir, root, "code");

  return { displayFile, codeDir };
}

export function getFixturesForProjectTest(
  {
    excludeFixtures,
    includeFixtures,
  }: {
    excludeFixtures?: Set<string> | string[];
    includeFixtures?: Set<string> | string[];
  } = { excludeFixtures: new Set(), includeFixtures: new Set() }
): {
  fixtureName: string;
  displayFile: string;
  codeDir: string;
}[] {
  return getFixtures(excludeFixtures, includeFixtures).map((root) => ({
    fixtureName: root,
    ...getFixturePaths(root),
  }));
}

export function getAllFilesInDir(
  dirPath: string,
  ignoreRegexes?: RegExp[]
): { [relativePath: string]: string } {
  const files: { [relativePath: string]: string } = {};

  function traverseDir(currentPath: string) {
    const entries = readdirSync(currentPath, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = join(currentPath, entry.name);

      if (entry.isDirectory()) {
        traverseDir(fullPath);
      } else if (entry.isFile()) {
        const relativePath = relative(dirPath, fullPath);

        if (
          ignoreRegexes &&
          ignoreRegexes.some((regex) => regex.test(relativePath))
        ) {
          continue;
        }

        files[relativePath] = fullPath;
      }
    }
  }

  traverseDir(dirPath);
  return files;
}
