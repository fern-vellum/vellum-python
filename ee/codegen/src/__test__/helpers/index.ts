export { nodeContextFactory } from "./node-context-factory";
export { workflowContextFactory } from "./workflow-context-factory";
