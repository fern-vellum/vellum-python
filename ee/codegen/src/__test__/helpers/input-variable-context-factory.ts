import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { InputVariableContext } from "src/context/input-variable-context";

export function inputVariableContextFactory({
  inputVariableData,
  workflowContext,
}: {
  inputVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): InputVariableContext {
  return new InputVariableContext({
    inputVariableData,
    workflowContext,
  });
}
