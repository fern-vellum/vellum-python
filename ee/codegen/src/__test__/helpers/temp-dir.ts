import os from "os";
import path from "path";

export function makeTempDir(base?: string): string {
  const randomSuffix = Math.random().toString(36).substring(7);
  return path.join(os.tmpdir(), `${base ?? "test-directory"}-${randomSuffix}`);
}
