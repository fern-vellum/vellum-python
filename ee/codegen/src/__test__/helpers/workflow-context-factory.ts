import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowLabel,
  workflowClassName,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowLabel || "Workflow",
    workflowLabel: workflowClassName || "Workflow",
    vellumApiKey: "<TEST_API_KEY>",
  });
}
