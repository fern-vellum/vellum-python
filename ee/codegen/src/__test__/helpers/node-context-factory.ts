import { workflowContextFactory } from "./workflow-context-factory";

import { createNodeContext } from "src/context";
import { BaseNodeContext } from "src/context/node-context/base";
import {
  SearchNode,
  WorkflowDataNode,
  WorkflowNodeType as WorkflowNodeTypeEnum,
} from "src/types/vellum";

export async function nodeContextFactory({
  workflowContext,
  nodeData,
}: Partial<BaseNodeContext.Args<WorkflowDataNode>> = {}): Promise<
  BaseNodeContext<WorkflowDataNode>
> {
  return await createNodeContext({
    workflowContext: workflowContext ?? workflowContextFactory(),
    nodeData: nodeData ?? {
      id: "search",
      type: WorkflowNodeTypeEnum.SEARCH,
      inputs: [],
      data: { label: "MyNode" } as SearchNode["data"],
    },
  });
}
