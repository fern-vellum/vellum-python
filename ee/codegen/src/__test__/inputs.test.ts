import { Writer } from "@fern-api/python-ast/core/Writer";
import { VellumVariable } from "vellum-ai/api";

import { workflowContextFactory } from "./helpers";
import { inputVariableContextFactory } from "./helpers/input-variable-context-factory";

import * as codegen from "src/codegen";
import { WorkflowContext } from "src/context";

describe("Inputs", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();
  });

  describe("write", () => {
    it("should generate correct code when Inputs has no variables", async () => {
      const inputs = codegen.inputs({ workflowContext });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code when Inputs has variables", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "input1", key: "input1", type: "STRING" },
        { id: "input2", key: "input2", type: "NUMBER" },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({ workflowContext });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should generate correct code when Inputs has a custom name", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "input1", key: "input1", type: "STRING" },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({
        workflowContext,
        name: "CustomInputs",
      });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should not generate any code when Inputs is empty", async () => {
      const inputs = codegen.inputs({ workflowContext });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toBe("");
    });

    it("should generate correct code for complex input variables", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "1", key: "query", type: "STRING" },
        { id: "2", key: "max_runtime", type: "NUMBER" },
        { id: "3", key: "previous_chat_history", type: "CHAT_HISTORY" },
        { id: "4", key: "prior_results", type: "SEARCH_RESULTS" },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({
        workflowContext,
      });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("should convert input variable names into valid python attributes", async () => {
      const inputVariables: VellumVariable[] = [
        { id: "1", key: "My Input", type: "STRING" },
        { id: "2", key: "$My*Input", type: "NUMBER" },
      ];
      inputVariables.forEach((inputVariableData) => {
        workflowContext.addInputVariableContext(
          inputVariableContextFactory({
            inputVariableData: inputVariableData,
            workflowContext,
          })
        );
      });
      const inputs = codegen.inputs({
        workflowContext,
      });

      inputs.write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
