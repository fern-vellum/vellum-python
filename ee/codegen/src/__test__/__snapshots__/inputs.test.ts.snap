// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Inputs > write > should generate correct code for complex input variables 1`] = `
"from vellum.workflows.inputs import BaseInputs
from typing import List
from vellum import ChatMessage, SearchResult


class Inputs(BaseInputs):
    query: str
    max_runtime: float
    previous_chat_history: List[ChatMessage]
    prior_results: List[SearchResult]
"
`;

exports[`Inputs > write > should generate correct code when Inputs has a custom name 1`] = `
"from vellum.workflows.inputs import BaseInputs


class CustomInputs(BaseInputs):
    input1: str
"
`;

exports[`Inputs > write > should generate correct code when Inputs has no variables 1`] = `""`;

exports[`Inputs > write > should generate correct code when Inputs has variables 1`] = `
"from vellum.workflows.inputs import BaseInputs


class Inputs(BaseInputs):
    input1: str
    input2: float
"
`;
